                        Aum Sri Sairam
                                                    Date:20/12/2017
# Linear Regression using boston dataset


We will implement linear regression using boston dataset available in sklearn.datasets. To know the how perfect how model performs we will use the <i>mean_square_error</i> (available in sklearn). We will be fine tuning our model using parameters alpha(i.e learning rate) and epsilon(error tolerance). At the end, to see the performance of our model we will try to compute the test error and training error. Any suggestion that will contribute in the improvement of this program is most welcome.

In [376]:
from sklearn.datasets import load_boston
from sklearn.metrics import accuracy_score, mean_squared_error
from sklearn.model_selection import train_test_split
import numpy as np

In [377]:
boston= load_boston()
X= boston.data
y= boston.target

Insert a columns of 1 to x

In [378]:
X= np.insert(X,0,1,axis=1)

In [379]:
theta = np.ones(X[0].shape)

First, I declare some parameters: alpha is my learning rate, and epsilon is 2-norm of difference between the two consecutive thetas. I will be using np.dot for inner product. If norm values is than epsilon, it means that theta and oldTheta are close to each other,in other words, theta has converged.

In [414]:
alpha= 1e-6
epsilon= 1e-7

Now define a cost function that computes the cost of using theta as the parameter for linear regression to fit the data points in X and y.<br>
## Notation:
X: denotes a matrix<br>
y:vector<br>
Note: dimension of theta has to be same as the dimension of each x<sub>i</sub> + 1

In [415]:
def costFunction(X,y,theta):
    m= len(y)
    J= np.sum((X.dot(theta) - y)**2)/(2*m)
    return J

Now define a gradientDescent function that performs gradient descent to learn theta.

In [416]:
def gradientDescent(X,y,theta,alpha):
    oldTheta= theta * 99
    m= len(y)
    norm= np.linalg.norm(theta - oldTheta , ord=2)
    iteration= 0
    costHistory= []
    while norm >= epsilon:
        #print"  norm: ", norm
        hypothesis= X.dot(theta)
        loss= hypothesis - y
        gradient= X.T.dot(loss) / m
        oldTheta= theta
        theta= theta - alpha * gradient
        norm= np.linalg.norm(theta - oldTheta , ord=2)
       # print"gradient", gradient
        cost= costFunction(X,y,theta)
        costHistory.append(cost)
    return theta, costHistory

In [417]:
X_train, X_test, y_train, y_test= train_test_split(X,y,test_size=0.33,random_state=42)

In [418]:
optTheta, cost= gradientDescent(X_train,y_train,theta,alpha)
#print cost

In [419]:
y_predict= X_train.dot(optTheta)

In [420]:
error= mean_squared_error(y_predict, y_train)
print error

24.9660983814


In [421]:
 y_predict_test= X_test.dot(optTheta)
error= mean_squared_error(y_predict_test, y_test)
print error

23.5188965452


# PERFORMANCE ANALYSIS

|Error tolerance|learning rate|trainig MSE  | test MSE |
|---------------|-------------|-------------|----------|
|      0.00001  |    1e-6     |  46.47      | 41.74    |
|     0.000001  |    1e-6     |  26.55      | 25.67    |
|    0.0000001  |    1e-6     |  24.96      | 23.52    |

# Conclusion
From above analysis we see that as error tolerance decreases the MSE too decreases. The model performs better with epsilon= 0.0000001 and alpha= 0.000001.